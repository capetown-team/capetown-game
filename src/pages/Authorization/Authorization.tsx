import React, {
  useEffect,
  useState,
  MouseEvent,
  FocusEvent,
  ChangeEvent,
  useCallback
} from 'react';
import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import { useHistory, Link } from 'react-router-dom';
import block from 'bem-cn-lite';

import { AppState } from '@/reducers';
import { signIn } from '@/reducers/user/actions';
import {
  authSelector,
  errorSelector,
  loadSelector
} from '@/reducers/user/selectors';
import { Button } from '@/components/Button';
import { Input } from '@/components/Input';
import { Loading } from '@/components/Loading';
import { isValidLogin, isValidPassword } from '@/modules/validation';
import { ROUTES } from '@/constants';
import { getCodeOAuth } from '@/modules/OAuth';

import './Authorization.scss';

const b = block('form');

export const Authorization = () => {
  const dispatch = useDispatch();
  const history = useHistory();

  const [loading, setLoading] = useState(true);
  const [login, setLogin] = useState('');
  const [password, setPassword] = useState('');
  const [loginDirty, setLoginDirty] = useState(false);
  const [passwordDirty, setPasswordDirty] = useState(false);
  const [loginError, setLoginError] = useState('Email не может быть пустым');
  const [passwordError, setPasswordError] = useState(
    'Пароль не может быть пустым'
  );
  const [formValid, setFormValid] = useState(false);
  const [regValid, setRegValid] = useState(true);

  const { isAuth, error, pending } = useSelector((state: AppState) => {
    return {
      isAuth: authSelector(state),
      pending: loadSelector(state),
      error: errorSelector(state)
    };
  }, shallowEqual);

  useEffect(() => {
    setLoading(pending);
  }, [pending]);

  useEffect(() => {
    if (loginError || passwordError) {
      setFormValid(false);
    } else {
      setFormValid(true);
    }
  }, [loginError, passwordError]);

  useEffect(() => {
    if (isAuth) {
      history.replace(ROUTES.GAME);
    }
  }, [isAuth, history]);

  useEffect(() => {
    setRegValid(!error);
  }, [error]);

  const blurHandler = useCallback((e: FocusEvent<Element>) => {
    const { name } = e.target as HTMLInputElement;
    switch (name) {
      case 'login':
        setLoginDirty(true);
        break;
      case 'password':
        setPasswordDirty(true);
        break;
      default:
    }
  }, []);

  const loginHandler = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    const text = e.target.value;
    setLogin(text);
    const loginErr = isValidLogin(text);
    setLoginError(loginErr);
  }, []);

  const passwordHandler = useCallback((e: ChangeEvent<HTMLInputElement>) => {
    const text = e.target.value;
    setPassword(text);
    const passwordErr = isValidPassword(text);
    setPasswordError(passwordErr);
  }, []);

  const submitHandler = useCallback(
    (e: MouseEvent<Element>) => {
      e.preventDefault();

      const user = {
        login,
        password
      };

      dispatch(signIn(user));
    },
    [dispatch, login, password]
  );

  const oAuthHandler = () => {
    getCodeOAuth();
  };

  return (
    <div className={b()}>
      {loading && <Loading />}
      <form className={b('wrapper')}>
        <div className={b('title')}>Авторизация</div>
        <div className={b('main')}>
          <div className={b('row')}>
            <div className={b('title-input')}> Логин</div>
            {loginDirty && loginError && (
              <div style={{ color: 'red' }}>{loginError}</div>
            )}
            <Input
              id="login"
              value={login}
              name="login"
              onChange={loginHandler}
              onBlur={(e) => blurHandler(e)}
              placeholder="Логин"
            />
          </div>
          <div className={b('row')}>
            <div className={b('title-input')}> Пароль </div>
            {passwordDirty && passwordError && (
              <div style={{ color: 'red' }}>{passwordError}</div>
            )}
            <Input
              id="password"
              value={password}
              type="password"
              name="password"
              onChange={passwordHandler}
              onBlur={(e) => blurHandler(e)}
              placeholder="Пароль"
            />
          </div>
          {!regValid && (
            <div style={{ color: 'red' }}>Логин или пароль не верный</div>
          )}
          <div className={b('row-button')}>
            <Button
              disabled={!formValid}
              type="submit"
              size="m"
              onClick={(e: React.MouseEvent<Element, globalThis.MouseEvent>) =>
                submitHandler(e)
              }
            >
              Вход
            </Button>
          </div>
          <Link className={b('link')} to={ROUTES.SIGNUP}>
            Регистрация
          </Link>
          <div className={b('row-button-oauth')}>
            <Button size="m" onClick={() => oAuthHandler()}>
              Войти через Yandex
            </Button>
          </div>
        </div>
      </form>
    </div>
  );
};
