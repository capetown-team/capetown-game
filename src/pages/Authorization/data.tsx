import { FormDataType } from '@/types.d';

export const data: FormDataType = {
  login: {
    element: 'input',
    value: '',
    config: {
      name: 'name_input',
      type: 'text',
      placeholder: 'Введите ваше имя',
      label: 'Имя'
    },
    validation: {
      required: true,
      minLength: 4
    },
    valid: false,
    touched: false,
    validationMessage: ''
  },
  password: {
    element: 'input',
    value: '',
    config: {
      name: 'password_input',
      type: 'password',
      placeholder: 'Введите пароль',
      label: 'Пароль'
    },
    validation: {
      required: true,
      minLength: 8,
      password: true
    },
    valid: false,
    touched: false,
    validationMessage: ''
  }
};
