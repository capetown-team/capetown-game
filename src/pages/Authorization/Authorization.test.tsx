import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { Authorization } from './Authorization';

describe('<Authorization />', () => {
  it('should renders correct <Authorization />', () => {
    const wrapper = shallow(<Authorization />);

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
