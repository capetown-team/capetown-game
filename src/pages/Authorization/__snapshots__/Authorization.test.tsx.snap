// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Authorization /> should renders correct <Authorization /> 1`] = `
<div
  className="form__wrapper"
>
  <form
    className="form__login"
  >
    <div
      className="form__title"
    >
      Авторизация
    </div>
    <div
      className="form__main"
    >
      <div
        className="form__row"
      >
        <div
          className="form__title-input"
        >
           Логин
        </div>
        <Input
          id="login"
          name="login"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Логин"
          value=""
        />
      </div>
      <div
        className="row"
      >
        <div
          className="form__title-input"
        >
           Пароль 
        </div>
        <Input
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Пароль"
          type="password"
          value=""
        />
      </div>
      <div
        className="form__row-button"
      >
        <Memo(Button)
          disabled={true}
          onClick={[Function]}
          size="s"
          type="submit"
        >
          Вход
        </Memo(Button)>
      </div>
      <div
        className="form__row"
      >
        <Link
          className="form__link"
          to="/registration"
        >
          Регистрация
        </Link>
      </div>
    </div>
  </form>
</div>
`;
